### PR TITLE
Refactor MapAndBatchDatasetOp

### PR DESCRIPTION
**Pull request by feihugis**
This PR refactors `MapAndBatchDatasetOp` and adds the tests.

cc: @jsimsa

### CODE REVIEW

**Comment by jsimsa** [comment]
@rachellim please take a look

**Review by rachellim — COMMENTED** [review]
lgtm!

**Comment by rachellim** [comment]
Thanks for the change, @feihugis

File: tensorflow/core/kernels/data/experimental/BUILD
Comment (feihugis) [reply tensorflow/core/kernels/data/experimental/BUILD]: Yeah, will remove it.

**Comment by feihugis** [comment]
Thanks for the review, @rachellim! We are refactoring the test style to avoid the repeated code. This PR will be updated accordingly. Will ping you once I finish the change.

**Comment by feihugis** [comment]
We need to wait for #31592 to be merged. Once merged, the failures in `Ubuntu CC` will disappear.

**Comment by feihugis** [comment]
@rachellim The refactoring of the test style is good for review. Could you please take a look at the change(https://github.com/tensorflow/tensorflow/pull/30971/commits/18cb0f08d873fbae678f0115e850bd3e78fd5f8c) when you have time?

**Review by rachellim — DISMISSED** [review]
Thanks for the change! This lgtm.

**Comment by feihugis** [comment]
@rachellim The internal checks failed. Could you please help check the logs and paste them here? Thanks!

**Comment by rachellim** [comment]
`MapAndBatchDatasetOpTest.InvalidArguments` is giving SIGSEV. Trace:

```
PC: @     0x7fe2e28a51b4  (unknown)  tensorflow::UniqueTensorReferences::~UniqueTensorReferences()
    @     0x7fe2dba65d51       1648  FailureSignalHandler()
    @     0x7fe2db2649a0  245665408  (unknown)
    @     0x7fe2e283ee99        144  tensorflow::OpKernelContext::~OpKernelContext()
    @     0x7fe2e44ea939        224  tensorflow::data::DatasetOpsTestBase::CreateOpKernelContext()
    @     0x7fe2e44ecaa8         64  tensorflow::data::DatasetOpsTestBase::CreateDatasetContext()
    @     0x7fe2e4523306        112  tensorflow::data::experimental::(anonymous namespace)::MapAndBatchDatasetOpTest::Initialize()
    @     0x7fe2e45324ec       2224  tensorflow::data::experimental::(anonymous namespace)::MapAndBatchDatasetOpTest_InvalidArguments_Test::TestBody()
    @     0x7fe2e2bd392a         48  testing::Test::Run()
    @     0x7fe2e2bd4ab0         96  testing::TestInfo::Run()
    @     0x7fe2e2bd54a7         80  testing::TestSuite::Run()
    @     0x7fe2e2be4df7        208  testing::internal::UnitTestImpl::RunAllTests()
    @     0x7fe2e2be4415         64  testing::UnitTest::Run()
    @     0x7fe2e450336e         32  main
    @     0x7fe2db0d2bbd        208  __libc_start_main
    @     0x5645316cf4a9  (unknown)  ../sysdeps/x86_64/start.S:108 _start
```

**Comment by feihugis** [comment]
@rachellim Thanks for the detailed logs! The SIGSEV issue is addressed here (https://github.com/tensorflow/tensorflow/pull/30971/commits/97ae500d64f64b4ed056a2f7ea5e333ed7408620). Could you please take another look?

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Review by rachellim — APPROVED** [review]

**Pull request merged**
